File: frontend/src/features/audio/inputProfile.ts
/**
 * 入力モードと入力方式別プリセット（0019）。
 *
 * Backend `backend/services/input_profile.py` と同じ表を持つ。
 * 片方だけ変えると UI 表示と実処理がずれるため、必ず両方を更新すること。
 * 実際の音声処理は Backend が行い、ここは「何を送るか」と「何を見せるか」を決める。
 */

export type InputMode = 'auto' | 'mic' | 'loopback' | 'custom';
/** `auto` を解決したあとの実効モード。 */
export type ResolvedMode = 'mic' | 'loopback' | 'custom';
export type GainMode = 'auto' | 'none' | 'manual';
export type SilenceMode = 'relative' | 'absolute' | 'manual';

export const INPUT_MODES: readonly InputMode[] = ['auto', 'mic', 'loopback', 'custom'];
export const GAIN_MODES: readonly GainMode[] = ['auto', 'none', 'manual'];
export const SILENCE_MODES: readonly SilenceMode[] = ['relative', 'absolute', 'manual'];

/** 設計値。Backend の services/audio_levels.py と一致させる。 */
export const MAX_GAIN_DB = 24;
export const TARGET_SPEECH_DBFS = -24;
export const ABSOLUTE_SILENCE_RMS = 0.0002;
export const MIN_WARNING_SECONDS = 5;
export const MAX_WARNING_SECONDS = 120;
export const DEFAULT_WARNING_SECONDS = 20;

export const INPUT_MODE_LABELS: Record<InputMode, string> = {
  auto: '自動',
  mic: 'マイク',
  loopback: '内部音声',
  custom: 'カスタム'
};

export const INPUT_MODE_HINTS: Record<InputMode, string> = {
  auto: 'デバイス名から自動で判定します',
  mic: '遠い話者に合わせて音量を自動補正します',
  loopback: 'BlackHole 等。音声を一切加工しません',
  custom: '補正量や無音判定を手動で指定します'
};

/**
 * 仮想オーディオデバイス（ループバック）の既知の名前。
 * Backend の LOOPBACK_PATTERNS と同じ。
 */
const LOOPBACK_PATTERNS = [
  /blackhole/i,
  /soundflower/i,
  /loopback\s*audio/i,
  /vb[-\s]?cable/i,
  /vb[-\s]?audio/i,
  /existential\s*audio/i,
  /ishowu/i,
  /aggregate\s*device/i,
  /multi[-\s]?output/i,
  /virtual\s*(audio|cable|input)/i
];

/**
 * デバイス名から入力モードを自動判定する。
 *
 * 明確に判定できる仮想オーディオデバイスだけ `loopback`。
 * それ以外はすべて `mic`（判定に自信が無いときにマイク扱いにしておけば、
 * 補正が効くだけで音は壊れない）。
 */
export function detectMode(deviceLabel: string | null | undefined): 'mic' | 'loopback' {
  const label = String(deviceLabel ?? '').trim();
  if (!label) return 'mic';
  return LOOPBACK_PATTERNS.some((pattern) => pattern.test(label)) ? 'loopback' : 'mic';
}

export interface ResolvedInputMode {
  /** 実際に使うモード。 */
  mode: ResolvedMode;
  /** 自動判定の結果。UI へ必ず表示する。 */
  detected: 'mic' | 'loopback';
  /** 設定に保存されている選択。 */
  requested: InputMode;
  /** 自動判定をそのまま使っているか。 */
  isAuto: boolean;
}

/** 保存済みの選択とデバイス名から実効モードを決める。 */
export function resolveMode(
  requested: InputMode | string | null | undefined,
  deviceLabel: string | null | undefined
): ResolvedInputMode {
  const detected = detectMode(deviceLabel);
  const raw = String(requested ?? 'auto') as InputMode;
  const normalized: InputMode = (INPUT_MODES as readonly string[]).includes(raw) ? raw : 'auto';
  if (normalized === 'auto') {
    return { mode: detected, detected, requested: 'auto', isAuto: true };
  }
  return { mode: normalized as ResolvedMode, detected, requested: normalized, isAuto: false };
}

/** 1 デバイス分の入力設定。 */
export interface InputProfileSettings {
  inputMode: InputMode;
  gainMode: GainMode;
  manualGainDb: number;
  maxGainDb: number;
  silenceMode: SilenceMode;
  manualSilenceRms: number;
  lowInputWarning: boolean;
  lowInputWarningSeconds: number;
}

/** 入力方式別プリセット。値の根拠は docs/issues/0019。 */
export const PRESETS: Record<'mic' | 'loopback', Omit<InputProfileSettings, 'inputMode'>> = {
  mic: {
    gainMode: 'auto',
    manualGainDb: 0,
    maxGainDb: MAX_GAIN_DB,
    silenceMode: 'relative',
    manualSilenceRms: ABSOLUTE_SILENCE_RMS,
    lowInputWarning: true,
    lowInputWarningSeconds: DEFAULT_WARNING_SECONDS
  },
  loopback: {
    // 既にアプリ側で正規化された音が来るので触らない。
    // 無音が厳密なデジタル 0 になりノイズフロアが 0 へ張り付くため、絶対判定のみ。
    gainMode: 'none',
    manualGainDb: 0,
    maxGainDb: 0,
    silenceMode: 'absolute',
    manualSilenceRms: ABSOLUTE_SILENCE_RMS,
    lowInputWarning: true,
    lowInputWarningSeconds: DEFAULT_WARNING_SECONDS
  }
};

export const DEFAULT_PROFILE: InputProfileSettings = {
  inputMode: 'auto',
  ...PRESETS.mic
};

function clampNumber(value: unknown, fallback: number, min: number, max: number): number {
  const parsed = typeof value === 'number' ? value : Number(value);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.min(max, Math.max(min, parsed));
}

function pick<T extends string>(value: unknown, allowed: readonly T[], fallback: T): T {
  const raw = String(value ?? '');
  return (allowed as readonly string[]).includes(raw) ? (raw as T) : fallback;
}

/**
 * 保存値を安全な設定へ正規化する。
 *
 * - 未設定は既定値で埋める
 * - 不正値・範囲外は安全な値へ落とす
 * - `custom` 以外はプリセットを優先する
 *   （モードを選び直したときに前のモードの値が残らないようにする）
 * - 警告の ON/OFF と待ち時間だけはモードを問わずユーザー設定を尊重する
 */
export function normalizeProfile(
  raw: Partial<InputProfileSettings> | Record<string, unknown> | null | undefined,
  deviceLabel: string | null | undefined
): InputProfileSettings {
  const source = (raw ?? {}) as Record<string, unknown>;
  const inputMode = pick<InputMode>(source.inputMode, INPUT_MODES, 'auto');
  const { mode } = resolveMode(inputMode, deviceLabel);

  const warningSeconds = clampNumber(
    source.lowInputWarningSeconds,
    DEFAULT_WARNING_SECONDS,
    MIN_WARNING_SECONDS,
    MAX_WARNING_SECONDS
  );
  const lowInputWarning =
    typeof source.lowInputWarning === 'boolean' ? source.lowInputWarning : true;

  if (mode === 'custom') {
    return {
      inputMode,
      gainMode: pick<GainMode>(source.gainMode, GAIN_MODES, 'auto'),
      manualGainDb: clampNumber(source.manualGainDb, 0, 0, MAX_GAIN_DB),
      maxGainDb: clampNumber(source.maxGainDb, MAX_GAIN_DB, 0, MAX_GAIN_DB),
      silenceMode: pick<SilenceMode>(source.silenceMode, SILENCE_MODES, 'relative'),
      manualSilenceRms: clampNumber(source.manualSilenceRms, ABSOLUTE_SILENCE_RMS, 0, 0.05),
      lowInputWarning,
      lowInputWarningSeconds: warningSeconds
    };
  }

  const preset = PRESETS[mode];
  return {
    inputMode,
    ...preset,
    lowInputWarning,
    lowInputWarningSeconds: warningSeconds
  };
}

/** デバイスラベル -> 設定 の保存。ラベルが安定識別子（deviceId は origin で変わる）。 */
export type InputProfileMap = Record<string, Partial<InputProfileSettings>>;

/** 保存された map を安全な形へ整える。壊れた値は捨てるが、未知のキーは残さない。 */
export function normalizeProfileMap(raw: unknown): InputProfileMap {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return {};
  const out: InputProfileMap = {};
  for (const [label, value] of Object.entries(raw as Record<string, unknown>)) {
    if (!label || typeof value !== 'object' || value === null || Array.isArray(value)) continue;
    out[label] = value as Partial<InputProfileSettings>;
  }
  return out;
}

/**
 * デバイスに対応する設定を解決する。
 *
 * **フォールバック時に旧デバイスの設定を引き継がない。**
 * 保存が無ければ、そのデバイス名で判定したプリセットを使う。
 * BlackHole 用の「補正なし」を本体マイクへ持ち込むと 0019 が再発する。
 */
export function profileForDevice(
  profiles: InputProfileMap,
  deviceLabel: string | null | undefined,
  fallback?: Partial<InputProfileSettings>
): InputProfileSettings {
  const label = String(deviceLabel ?? '').trim();
  const saved = label ? profiles[label] : undefined;
  if (saved) return normalizeProfile(saved, label);
  // 保存が無いデバイスでは、渡された既定ではなく必ずデバイス判定から作り直す。
  const seed = fallback && !label ? fallback : { inputMode: 'auto' as InputMode };
  return normalizeProfile(seed, label);
}

/** デバイス別設定を書き戻す。既存の他デバイスの設定は触らない。 */
export function rememberProfile(
  profiles: InputProfileMap,
  deviceLabel: string | null | undefined,
  settings: InputProfileSettings
): InputProfileMap {
  const label = String(deviceLabel ?? '').trim();
  if (!label) return profiles;
  return { ...profiles, [label]: { ...settings } };
}

export interface ActualDeviceResolution {
  settings: InputProfileSettings;
  /** 実デバイスが保存済みラベルと異なり、プリセットを解決し直したか。 */
  relabeled: boolean;
  /** 実際に使うデバイス名。Backend の自動判定はこの名前で行わせる。 */
  deviceLabel: string;
}

/**
 * **録音開始直前に、実際に開けたデバイスでプリセットを解決し直す（0019）。**
 *
 * 設定は保存済みデバイス名で読み込むが、`getUserMedia` が実際に開くデバイスは
 * フォールバックで別物になりうる（USB の抜き差し、origin 変更、既定入力への後退）。
 * このとき保存済み設定をそのまま使うと、たとえば BlackHole 用の
 * 「補正なし・絶対無音判定」が本体マイクへ適用され、0019 の欠落条件が再現する。
 *
 * 実デバイス名が保存済みと異なるときは、保存済み設定を捨て、
 * **実デバイス名で `inputProfiles` を引き直す**（無ければデバイス名から判定し直す）。
 */
export function resolveProfileForActualDevice(
  savedLabel: string | null | undefined,
  actualLabel: string | null | undefined,
  savedProfile: InputProfileSettings,
  profiles: InputProfileMap
): ActualDeviceResolution {
  const saved = String(savedLabel ?? '').trim();
  const actual = String(actualLabel ?? '').trim();

  // 実デバイス名が取れない（権限やブラウザの都合でラベルが空）ときは、
  // 判断材料が無いので保存済み設定をそのまま使う。捨てる側へ倒さない。
  if (!actual || actual === saved) {
    return { settings: savedProfile, relabeled: false, deviceLabel: actual || saved };
  }
  return {
    settings: profileForDevice(profiles, actual),
    relabeled: true,
    deviceLabel: actual
  };
}

/** Backend の `input_profile` payload。WebSocket の config へそのまま載せる。 */
export function toBackendPayload(settings: InputProfileSettings, deviceLabel: string): {
  mode: InputMode;
  device_label: string;
  gain_mode: GainMode;
  manual_gain_db: number;
  max_gain_db: number;
  silence_mode: SilenceMode;
  manual_silence_rms: number;
  low_input_warning: boolean;
  low_input_warning_seconds: number;
} {
  return {
    mode: settings.inputMode,
    device_label: deviceLabel,
    gain_mode: settings.gainMode,
    manual_gain_db: settings.manualGainDb,
    max_gain_db: settings.maxGainDb,
    silence_mode: settings.silenceMode,
    manual_silence_rms: settings.manualSilenceRms,
    low_input_warning: settings.lowInputWarning,
    low_input_warning_seconds: settings.lowInputWarningSeconds
  };
}

/** 自動判定の結果を UI に出す短い説明。 */
export function describeResolution(resolution: ResolvedInputMode, deviceLabel: string): string {
  const device = deviceLabel.trim() || '既定の入力デバイス';
  if (!resolution.isAuto) {
    return `${device} → ${INPUT_MODE_LABELS[resolution.requested]}（手動で選択中）`;
  }
  return `${device} → ${INPUT_MODE_LABELS[resolution.detected]}（自動判定）`;
}

File: backend/services/input_profile.py
"""入力モードと入力方式別プリセット（0019）。

デバイス名だけで処理を固定しない。「入力モード」と「デバイス別設定」を組み合わせ、
自動判定の結果は UI へ出してユーザーが上書きできるようにする。

frontend/src/features/audio/inputProfile.ts と同じ表を持つ。
片方だけ変えると UI 表示と実処理がずれるため、必ず両方を更新すること。
"""
import re
from dataclasses import dataclass, asdict
from typing import Optional

from .audio_levels import (
    ABSOLUTE_SILENCE_RMS,
    MAX_GAIN_DB,
    SPEECH_OVER_FLOOR_DB,
    TARGET_SPEECH_DBFS,
)

# --- 入力モード ---------------------------------------------------------------

MODE_AUTO = "auto"
MODE_MIC = "mic"
MODE_LOOPBACK = "loopback"
MODE_CUSTOM = "custom"
INPUT_MODES = (MODE_AUTO, MODE_MIC, MODE_LOOPBACK, MODE_CUSTOM)

# --- ゲインモード -------------------------------------------------------------

GAIN_AUTO = "auto"
GAIN_NONE = "none"
GAIN_MANUAL = "manual"
GAIN_MODES = (GAIN_AUTO, GAIN_NONE, GAIN_MANUAL)

# --- 無音判定モード -----------------------------------------------------------

SILENCE_RELATIVE = "relative"
SILENCE_ABSOLUTE = "absolute"
SILENCE_MANUAL = "manual"
SILENCE_MODES = (SILENCE_RELATIVE, SILENCE_ABSOLUTE, SILENCE_MANUAL)

# 仮想オーディオデバイス（ループバック）の既知の名前。
# 単語境界で見るので "Microphone" のような無関係な語には当たらない。
LOOPBACK_PATTERNS = (
    r"blackhole",
    r"soundflower",
    r"loopback\s*audio",
    r"vb[-\s]?cable",
    r"vb[-\s]?audio",
    r"existential\s*audio",
    r"ishowu",
    r"aggregate\s*device",
    r"multi[-\s]?output",
    r"virtual\s*(audio|cable|input)",
)
_LOOPBACK_RE = re.compile("|".join(LOOPBACK_PATTERNS), re.IGNORECASE)

# 警告までの継続時間の許容範囲。
MIN_WARNING_SECONDS = 5.0
MAX_WARNING_SECONDS = 120.0
DEFAULT_WARNING_SECONDS = 20.0


def detect_mode(device_label: Optional[str]) -> str:
    """デバイス名から入力モードを自動判定する。

    明確に判定できる仮想オーディオデバイスだけ ``loopback``。
    それ以外はすべて ``mic``（判定に自信が無いときにマイク扱いにしておけば、
    補正が効くだけで音は壊れない）。
    """
    label = str(device_label or "").strip()
    if not label:
        return MODE_MIC
    return MODE_LOOPBACK if _LOOPBACK_RE.search(label) else MODE_MIC


def resolve_mode(requested_mode: Optional[str], device_label: Optional[str]) -> tuple[str, str]:
    """(実効モード, 自動判定結果) を返す。

    ``custom`` は実効モードとしてそのまま残す（プリセットではなくユーザー設定を使う）。
    ``auto`` のときだけ自動判定を実効モードに採用する。
    """
    detected = detect_mode(device_label)
    mode = str(requested_mode or MODE_AUTO).strip().lower()
    if mode not in INPUT_MODES:
        mode = MODE_AUTO
    if mode == MODE_AUTO:
        return detected, detected
    return mode, detected


def _clamp(value, default: float, minimum: float, maximum: float) -> float:
    try:
        parsed = float(value)
    except (TypeError, ValueError):
        return float(default)
    if parsed != parsed:  # NaN
        return float(default)
    return float(max(minimum, min(parsed, maximum)))


@dataclass
class InputProfile:
    """1 セッションで使う入力設定。プリセットとユーザー設定を解決した結果。"""

    mode: str = MODE_MIC
    detected_mode: str = MODE_MIC
    requested_mode: str = MODE_AUTO
    device_label: str = ""
    gain_mode: str = GAIN_AUTO
    manual_gain_db: float = 0.0
    max_gain_db: float = MAX_GAIN_DB
    target_speech_dbfs: float = TARGET_SPEECH_DBFS
    silence_mode: str = SILENCE_RELATIVE
    manual_silence_rms: float = ABSOLUTE_SILENCE_RMS
    speech_over_floor_db: float = SPEECH_OVER_FLOOR_DB
    clip_protection: bool = True
    low_input_warning: bool = True
    low_input_warning_seconds: float = DEFAULT_WARNING_SECONDS

    @property
    def gain_enabled(self) -> bool:
        """ゲインを 1 サンプルでも変更するか。``none`` かつ 0dB なら素通し。"""
        if self.gain_mode == GAIN_NONE:
            return False
        if self.gain_mode == GAIN_MANUAL:
            return abs(self.manual_gain_db) > 1e-9
        return self.max_gain_db > 0.0

    def as_dict(self) -> dict:
        data = asdict(self)
        data["gain_enabled"] = self.gain_enabled
        return data

    def summary(self) -> str:
        """diagnostics.log の 1 行。デバイス名は含めるがそれ以外の個体情報は出さない。"""
        return (
            f"input_profile mode={self.mode} requested={self.requested_mode} "
            f"detected={self.detected_mode} gain_mode={self.gain_mode} "
            f"max_gain_db={self.max_gain_db:.1f} manual_gain_db={self.manual_gain_db:.1f} "
            f"silence_mode={self.silence_mode} clip_protection={int(self.clip_protection)} "
            f"warn={int(self.low_input_warning)} warn_seconds={self.low_input_warning_seconds:.0f}"
        )


# --- プリセット ---------------------------------------------------------------
# 値の根拠は docs/issues/0019「入力方式別プリセット（初期値の根拠）」を参照。

PRESETS: dict[str, dict] = {
    MODE_MIC: {
        # 物理マイク。遠方の会話を拾うため自動補正を効かせる。
        "gain_mode": GAIN_AUTO,
        "manual_gain_db": 0.0,
        "max_gain_db": MAX_GAIN_DB,
        "silence_mode": SILENCE_RELATIVE,
        "manual_silence_rms": ABSOLUTE_SILENCE_RMS,
        "clip_protection": True,
        "low_input_warning": True,
        "low_input_warning_seconds": DEFAULT_WARNING_SECONDS,
    },
    MODE_LOOPBACK: {
        # 内部音声。既にアプリ側で正規化された音が来るので触らない。
        # 無音が厳密なデジタル 0 になりノイズフロアが 0 へ張り付くため、
        # 相対判定ではなく絶対下限だけで判定する。
        "gain_mode": GAIN_NONE,
        "manual_gain_db": 0.0,
        "max_gain_db": 0.0,
        "silence_mode": SILENCE_ABSOLUTE,
        "manual_silence_rms": ABSOLUTE_SILENCE_RMS,
        "clip_protection": True,
        "low_input_warning": True,
        "low_input_warning_seconds": DEFAULT_WARNING_SECONDS,
    },
}


def preset_for(mode: str) -> dict:
    """モードに対応する初期プリセット。``custom``/``auto`` は mic を土台にする。"""
    return dict(PRESETS.get(mode, PRESETS[MODE_MIC]))


def build_profile(payload: Optional[dict], device_label: Optional[str] = None) -> InputProfile:
    """設定 payload と デバイス名から実際に使うプロファイルを組み立てる。

    - 未知のキーは無視する（呼び出し側の設定ファイルは壊さない）
    - 不正値・欠損値は必ずプリセット値へフォールバックする
    - ``custom`` 以外では、ユーザーが個別に指定した値よりプリセットを優先する
      （モードを選び直したときに前のモードの値が残らないようにする）
    """
    data = payload if isinstance(payload, dict) else {}
    label = str(data.get("device_label") or device_label or "").strip()
    requested = str(data.get("mode") or MODE_AUTO).strip().lower()
    if requested not in INPUT_MODES:
        requested = MODE_AUTO
    mode, detected = resolve_mode(requested, label)

    base = preset_for(MODE_MIC if mode == MODE_CUSTOM else mode)

    if mode == MODE_CUSTOM:
        gain_mode = str(data.get("gain_mode") or base["gain_mode"]).strip().lower()
        if gain_mode not in GAIN_MODES:
            gain_mode = base["gain_mode"]
        silence_mode = str(data.get("silence_mode") or base["silence_mode"]).strip().lower()
        if silence_mode not in SILENCE_MODES:
            silence_mode = base["silence_mode"]
        max_gain_db = _clamp(data.get("max_gain_db"), base["max_gain_db"], 0.0, MAX_GAIN_DB)
        manual_gain_db = _clamp(data.get("manual_gain_db"), 0.0, 0.0, MAX_GAIN_DB)
        manual_silence_rms = _clamp(
            data.get("manual_silence_rms"), ABSOLUTE_SILENCE_RMS, 0.0, 0.05
        )
        low_input_warning = bool(data.get("low_input_warning", base["low_input_warning"]))
        warning_seconds = _clamp(
            data.get("low_input_warning_seconds"),
            base["low_input_warning_seconds"],
            MIN_WARNING_SECONDS,
            MAX_WARNING_SECONDS,
        )
    else:
        gain_mode = base["gain_mode"]
        silence_mode = base["silence_mode"]
        max_gain_db = float(base["max_gain_db"])
        manual_gain_db = float(base["manual_gain_db"])
        manual_silence_rms = float(base["manual_silence_rms"])
        # 警告の ON/OFF と待ち時間だけは、モードを問わずユーザー設定を尊重する
        # （うるさいと感じたら切れる、という操作を奪わない）。
        low_input_warning = bool(data.get("low_input_warning", base["low_input_warning"]))
        warning_seconds = _clamp(
            data.get("low_input_warning_seconds"),
            base["low_input_warning_seconds"],
            MIN_WARNING_SECONDS,
            MAX_WARNING_SECONDS,
        )

    return InputProfile(
        mode=mode,
        detected_mode=detected,
        requested_mode=requested,
        device_label=label,
        gain_mode=gain_mode,
        manual_gain_db=manual_gain_db,
        max_gain_db=max_gain_db,
        target_speech_dbfs=TARGET_SPEECH_DBFS,
        silence_mode=silence_mode,
        manual_silence_rms=manual_silence_rms,
        speech_over_floor_db=SPEECH_OVER_FLOOR_DB,
        # クリッピング防止は常に有効。切れるようにする理由が無い。
        clip_protection=True,
        low_input_warning=low_input_warning,
        low_input_warning_seconds=warning_seconds,
    )
